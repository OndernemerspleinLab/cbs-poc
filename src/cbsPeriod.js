import {addYears, addMonths, endOfYear, endOfMonth, endOfDay, format, parse} from 'date-fns';
import nlLocale from 'date-fns/locale/nl';

const formatDate = (date, formatTemplate) =>  format(date, formatTemplate, {locale: nlLocale});

const splitCbsPeriodString = (cbsPeriodString) => ({
    year: cbsPeriodString.slice(0, 4),
    part1: cbsPeriodString.slice(4, 6),
    part2: cbsPeriodString.slice(6, 8),
});

const createDate = ({
    year, month = 0, day = 1, hours = 0, minutes = 0, seconds = 0, milliseconds = 0,
}) => new Date(year, month, day, hours, minutes, seconds, milliseconds);

// Convert to zero based
const zeroBased = (index) => Number(index) - 1;
const regexMatcher = (regex) => ({part1}) => regex.test(part1);
const stringMatcher = (string) => ({part1}) => string === part1;

// All supported cbsPeriods
const cbsPeriodTypes = [
    {
        match: regexMatcher(/^\d\d$/),
        parse: ({year, part1, part2}) => {
            const startDate = createDate({year, month: zeroBased(part1), day: part2});
            const endDate = endOfDay(startDate);
            return {
                type: 'Dagen',
                startDate,
                endDate,
                format: () => formatDate(startDate, 'D MMM YYYY'),
            };
        },
    },
    {
        match: stringMatcher('JJ'),
        parse: ({year}) => {
            const startDate = createDate({year});
            const endDate = endOfYear(startDate);
            return {
                type: 'Jaar',
                startDate,
                endDate,
                format: () => formatDate(startDate, 'YYYY'),                
            };
        }
    },
    {
        match: stringMatcher('MM'),
        parse: ({year, part2}) => {
            const startDate = createDate({year, month: zeroBased(part2)});
            const endDate = endOfMonth(startDate);
            return {
                type: 'Maanden',
                startDate,
                endDate,
                format: () => formatDate(startDate, 'MMM YYYY'),                
            };
        },
    },
    {
        match: stringMatcher('KW'),
        parse: ({year, part2}) => {
            const startDate = createDate({year, month: zeroBased(part2) * 3});
            const endDate = endOfMonth(addMonths(startDate, 2)); // End of month adds the third month of the quarter
            return {
                type: 'Kwartalen',
                startDate,
                endDate,
                format: () => formatDate(startDate, 'YYYY [Q]Q'),                
            };
        }
    },
    {
        match: stringMatcher('SJ'),
        parse: ({year}) => {
            const startDate = createDate({year, month: 7, day: 1});
            const endDate = endOfMonth(addYears(createDate({year, month: 6}), 1));
            return {
                type: 'School-, Bouw-, Oogstjaar',
                startDate,
                endDate,
                format: () => `${formatDate(startDate, 'YYYY')}/${formatDate(endDate, 'YYYY')}`,                
            };
        },
    },
    {
        match: stringMatcher('HJ'),
        parse: ({year, part2}) => {
            const startMonth = zeroBased(part2) * 6;
            const startDate = createDate({year, month: startMonth});
            const endDate = endOfMonth(addMonths(startDate, 6));

            return {
                type: 'Half jaar',
                startDate,
                endDate,
                format: () => `${formatDate(startDate, 'YYYY')} ${part2 === '01' ? 'eerste helft' : 'tweede helft'}`,                
            };
        },
    },
    {
        match: stringMatcher('W1'),
        parse: ({year, part2}) => {
            return {
                type: 'Week, systeem 1',        
            };
        },
    },
    {
        match: stringMatcher('W4'),
        parse: ({year, part2}) => {
            return {
                type: 'Week, vier weken',        
            };
        },
    },
    {
        match: stringMatcher('VS'),
        parse: ({year, part2}) => {
            return {      
                type: 'Voorschijdende maanden',
            };
        },
    },
    {
        match: stringMatcher('G2'),
        parse: ({year, part2}) => {
            return {      
                type: '2-jaarsgemiddelde',
            };
        },
    },
    {
        match: stringMatcher('G3'),
        parse: ({year, part2}) => {
            return {      
                type: '3-jaarsgemiddelde',
            };
        },
    },
    {
        match: stringMatcher('G4'),
        parse: ({year, part2}) => {
            return {      
                type: '4-jaarsgemiddelde',
            };
        },
    },
    {
        match: stringMatcher('G5'),
        parse: ({year, part2}) => {
            return {      
                type: '5-jaarsgemiddelde',
            };
        },
    },
    {
        match: regexMatcher(/^X.$/),
        parse: ({year, part2}) => {
            return {      
                type: 'Geen officiÃ«le periode indeling',
            };
        },
    },
    {
        match: stringMatcher('VJ'),
        parse: ({year, part2}) => {
            return {      
                type: 'Voortschrijdend jaar',
            };
        },
    },
    {
        match: stringMatcher('M3'),
        parse: ({year, part2}) => {
            return {      
                type: '3 maandelijks gemiddelde',
            };
        },
    },
];

export const parseCbsPeriod = (cbsPeriodString) => {
    const cbsPeriod = splitCbsPeriodString(cbsPeriodString);

    const {parse} = cbsPeriodTypes.find(({match}) => match(cbsPeriod));

    return parse(cbsPeriod);
};