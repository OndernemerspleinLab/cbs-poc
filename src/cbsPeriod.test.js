import { parseCbsPeriod } from './cbsPeriod';

const snapShotTestCbsPeriodParsing = (description, cbsPeriodString) => {
    test(`CBS Period - ${description}`, () => {
        const { type, startDate, endDate, format } = parseCbsPeriod(cbsPeriodString);
        const formattedDate = format();
        const result = {
            input: cbsPeriodString,
            type,
            startDate: startDate.toString(),
            endDate: endDate.toString(),
            formattedDate,
        };
        expect(result).toMatchSnapshot();
    });
};

const cbsPeriodTests = [
    ['Dagen', '20150325'],
    ['Jaar', '2015JJ00'],
    ['Maanden', '2001MM06'],
    ['Kwartalen', '2017KW04'],
    ['School-, Bouw-, Oogstjaar', '2012SJ00'],
];

cbsPeriodTests.forEach((t) => snapShotTestCbsPeriodParsing(...t));