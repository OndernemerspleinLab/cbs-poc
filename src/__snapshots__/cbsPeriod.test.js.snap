exports[`test CBS Period - Dagen 1`] = `
Object {
  "endDate": "Wed Mar 25 2015 23:59:59 GMT+0100 (West-Europa (standaardtijd))",
  "formattedDate": "25 mar 2015",
  "input": "20150325",
  "startDate": "Wed Mar 25 2015 00:00:00 GMT+0100 (West-Europa (standaardtijd))",
  "type": "Dagen",
}
`;

exports[`test CBS Period - Jaar 1`] = `
Object {
  "endDate": "Thu Dec 31 2015 23:59:59 GMT+0100 (West-Europa (standaardtijd))",
  "formattedDate": "2015",
  "input": "2015JJ00",
  "startDate": "Thu Jan 01 2015 00:00:00 GMT+0100 (West-Europa (standaardtijd))",
  "type": "Jaar",
}
`;

exports[`test CBS Period - Kwartalen 1`] = `
Object {
  "endDate": "Sun Dec 31 2017 23:59:59 GMT+0100 (West-Europa (standaardtijd))",
  "formattedDate": "2017 Q4",
  "input": "2017KW04",
  "startDate": "Sun Oct 01 2017 00:00:00 GMT+0200 (West-Europa (zomertijd))",
  "type": "Kwartalen",
}
`;

exports[`test CBS Period - Maanden 1`] = `
Object {
  "endDate": "Sat Jun 30 2001 23:59:59 GMT+0200 (West-Europa (zomertijd))",
  "formattedDate": "jun 2001",
  "input": "2001MM06",
  "startDate": "Fri Jun 01 2001 00:00:00 GMT+0200 (West-Europa (zomertijd))",
  "type": "Maanden",
}
`;

exports[`test CBS Period - School-, Bouw-, Oogstjaar 1`] = `
Object {
  "endDate": "Wed Jul 31 2013 23:59:59 GMT+0200 (West-Europa (zomertijd))",
  "formattedDate": "2012/2013",
  "input": "2012SJ00",
  "startDate": "Wed Aug 01 2012 00:00:00 GMT+0200 (West-Europa (zomertijd))",
  "type": "School-, Bouw-, Oogstjaar",
}
`;
